Improved shower/blood effects

timothymtorres opened the pull request:
## About The Pull Request
Last night I was experimenting with hooking up different chemicals to a shower and discovered that blood didn't really do anything other than have red mist and particles. Your characters clothes were still cleaned. The mood boost was still happy. So I reworked it a bit.

Blood now:
- Gives a negative mood, disgust, and status effect when showering with it (unless you are morbid, evil, or undead, then it's considered positive)
- Has an icon alert for bloody showers
- Covers a mob's clothing with blood when showering (or any objects on the tile)
- Tossing or spraying a container full of blood now covers objects/mobs in blood
- The revenant defile spell now affects showers by removing all water recyclers and reagents that gets replaced with blood

Showers now:
- Require 70% of water to clean and get mood/status effects
- Require 70% of blood to get mood/status effects
- Require 20% of radioactive reagents to stop radiation removal effects

So it's possible to have a clean water shower that is secretly radioactive. Since radioactive reagents do nothing on `TOUCH`, all this achieves is preventing the water from washing off the radiation.

I did have to refactor some of the reagent code to support method types for objects since I was experiencing hazmat issues when I was testing. Whenever I would inject blood from a syringe into a beaker, it would cover the beaker in blood on the outside. This would have been extremely hazardous for viruses. So I needed to make sure we are only applying it to the methods for `VAPOR|TOUCH`

Also improved the mood typecasting for owner to allow checking of mob biotypes. (so we can check `UNDEAD` for mood)

## Why It's Good For The Game
Blood effects and interactions are now more consistent. The code for objects is refactored to support method interactions with reagents. Evil/Morbid people now get some unique interactions that fit their theme. Last we get a cool new ability to let revenant's make their defiled areas something out of a horror movie.

## Changelog
:cl:
add: The revenant defile spell now affects showers by removing all water recyclers and reagents that gets replaced with blood.
add: Showering in clean water (+70%) results in positive mood/regen/stamina effects. It will wash off the mob.
add: Showering in dirty water results in negative mood effects and disgust. It will NOT wash off the mob.
add: Showering with radioactive reagents (+20%) results in the preventing the shower from washing off the radiation.
add: Showering in blood (+70%) results in severe negative mood effects and disgust.  (unless you are morbid, evil, or undead then it's considered positive) It will cover the mob in blood.
add: Water effects that interact with a mob from touch or vapor (showering/spray bottles/etc.) will now heal sleep, unconsciousness, confusion, drowsiness, jitters, dizziness, and drunkenness.
fix: Fix bloody showers not covering objects in blood.
fix: Tossing or spraying a container full of blood now covers objects/mobs in blood
fix: Fix wrong status effect for watery tile
image: Add new alert icons for bloody/dirty showers
code: Refactored some expose_obj reagent code to support method types.
code: Improved mood typecasting for owner to allow checking of mob biotypes.
/:cl:

timothymtorres committed:
Add bloody shower effects

timothymtorres committed:
Merge branch 'master' into blood-shower

timothymtorres committed:
Add cursed bloody showers to defile spell

timothymtorres committed:
Remove debug settings

timothymtorres committed:
Combine all shower mood events into one

timothymtorres committed:
Add bloody shower screen alerts

timothymtorres committed:
Fix mobs and objects be covered in blood when exposed

timothymtorres committed:
Refactor signal events and procs for reagent code

DaCoolBoss commented:
Do vamps/undead count as morbid? If not, include those too

necromanceranne commented on code/datums/status_effects/neutral.dm:
```diff
@@ -620,17 +620,39 @@
+			alert_type = /atom/movable/screen/alert/status_effect/shower_regen/bloody_like
+		else
+			alert_type  = /atom/movable/screen/alert/status_effect/shower_regen/bloody_dislike
+	else if(HAS_TRAIT(owner, TRAIT_WATER_HATER) && !HAS_TRAIT(owner, TRAIT_WATER_ADAPTATION))
```
So you can be a shower hater AND enjoy a bloody shower? This might need more fine tuning. Especially the regen/degen ticks in the next proc.

timothymtorres replied:
It specifically says `WATER` in the trait. I also didn't really touch regen/degen stuff since I just wanted it to be mostly flavor and mood events.

necromanceranne replied:
Oh I misread that if statement, nevermind.

The reason I mention the regen/degen proc is because it is checking for TRAIT_WATER_HATER but not for these other traits. The flavour here is that the former causes you to take a degen because you hate showering in water. But if you like a bloody shower, it doesn't then make sense to suffer a degen. Does that make more sense?

timothymtorres replied:
Changed it to skip regen if it's a bloody shower since I'm aiming for this to be only a flavor effect.

necromanceranne commented:
I'm not too fussed if someone gives the Evil quirk even a single upside, but we should be careful not to just turn this into Morbid 2.0. It's not really meant to have upsides, as the quirk is both neutral AND mostly to antagonize a small subset of other players to no benefit of your own.

Not that this really adds new functionality, given that literally anyone can shower, and this just adds another reagent you can shower with.

timothymtorres committed:
Undead mobtypes now enjoy bloody showers

timothymtorres committed:
Remove healing from bloody showers

necromanceranne commented on code/datums/status_effects/neutral.dm:
```diff
@@ -620,21 +620,46 @@
 /datum/status_effect/shower_regen/tick(seconds_between_ticks)
 	. = ..()
+	if(bloody_shower)
+		return
```
Now blood arbitrarily doesn't interact the same as all the other TOUCH chems used in a shower. I suppose just removing the interaction for blood technically works, but this is kind of a shitty solution.

timothymtorres replied:
The old behavior was that if you had ANY chemical (even if it was NOT water) the regen effects happen and you got a positive mood boost for showering in something like oil or acid.

What should the mood boost be for showering in 3u water, 3u oil, 3u acid, 1u blood? What if the blood is infected? Which alert should be used? What about regen effects? etc. This was all stuff I considered. I ultimately decided to just use a boolean solution. Does the water have blood or not? Doesn't matter if it's 99% or 0.001%.

I'm also not sure where you are seeing the other chems getting bypassed? It only applies to the mood/regen effects. All the other TOUCH chem interactions should get be getting triggered.

necromanceranne replied:
I'm not a maint, so you are welcome to ignore me if you want since I'm asking a lot out of you rather unfairly. But it doesn't quite sit right with me to just switch it off if there is ever blood in the shower. I would expect there to just be a degen effect if there is more blood than other reagents in the shower, unless the person is happy with a bloody shower. Namely, it's only a bloody shower if there is majority blood in the reagents overall.

It might require a deeper amount of refactoring, but I think it's entirely possible. Probably by getting the volume of blood and comparing it to the volume of total reagents, and if it is above half, it's a bloody shower.

timothymtorres committed:
Fix mood events typecasting

MrMelbert commented on code/datums/status_effects/neutral.dm:
```diff
@@ -620,21 +620,46 @@
+	/// Is this the shower raining blood?
+	var/bloody_shower = FALSE
+
+/datum/status_effect/shower_regen/on_creation(mob/living/new_owner, _bloody_shower = FALSE)
+	bloody_shower = _bloody_shower
```
Prefer `src.var` to using `_var`

timothymtorres replied:
Done

MrMelbert commented on code/game/objects/objs.dm:
```diff
@@ -220,7 +220,8 @@ GLOBAL_LIST_EMPTY(objects_by_id_tag)
 	for(var/reagent in reagents)
 		var/datum/reagent/R = reagent
-		. |= R.expose_obj(src, reagents[R])
+		var/reac_volume = reagents[R]
+		. |= R.expose_obj(src, methods, reac_volume, show_message)
```
```suggestion
		. |= R.expose_obj(src, methods, reagents[R], show_message)
```

timothymtorres replied:
I deliberately added that var so that we know that reac_volume is being passed into the arg since it's not apparent. I ran into the problem of getting mixed up with volume_modifier since it's not clear that reagents[R] is used for volume.

timothymtorres replied:
I also should point out we are renaming an ambiguous one letter var.

MrMelbert commented on code/game/objects/structures/shower.dm:
```diff
@@ -253,17 +253,19 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/shower, (-16))
 
 /obj/machinery/shower/proc/wash_atom(atom/target)
-	target.wash(CLEAN_RAD | CLEAN_WASH)
+	var/datum/reagent/blood/bloody_shower = locate(/datum/reagent/blood) in reagents.reagent_list
```
I think you should be using `has_reagent` or `get_reagent` instead of `locate`

timothymtorres replied:
Done as well. I took Annie's suggestions and calculated if the reagents are +70% determine if the shower is bloody or clean. Also if radioactive reagents are +20% the shower will not remove radiation. (so it's possible to have a sneaky radioactive/clean shower)

MrMelbert commented on code/game/turfs/turf.dm:
```diff
@@ -697,7 +697,8 @@ GLOBAL_LIST_EMPTY(station_turfs)
 	for(var/reagent in reagents)
 		var/datum/reagent/R = reagent
-		. |= R.expose_turf(src, reagents[R])
+		var/reac_volume = reagents[R]
+		. |= R.expose_turf(src, reac_volume)
```
Seems weird to change this for no reason
```suggestion
		. |= R.expose_turf(src, reagents[R])
```

MrMelbert commented on code/modules/mob/living/living_defense.dm:
```diff
@@ -681,7 +681,8 @@
 	for(var/reagent in reagents)
 		var/datum/reagent/R = reagent
-		. |= R.expose_mob(src, methods, reagents[R], show_message, touch_protection)
+		var/reac_volume = reagents[R]
+		. |= R.expose_mob(src, methods, reac_volume, show_message, touch_protection)
```
Same
```suggestion
		. |= R.expose_mob(src, methods, reagents[R], show_message, touch_protection)
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -127,6 +133,26 @@
+			for(var/datum/disease/virus in data["viruses"])
+				if(virus.spread_flags & DISEASE_SPREAD_CONTACT_FLUIDS)
+					viruses_to_add += virus
+			if(LAZYLEN(viruses_to_add))
```
```suggestion
			if(length(viruses_to_add))
```

timothymtorres committed:
Rework shower reagents to use % of volume

timothymtorres committed:
Fix revenant defile ability not deleting all water

timothymtorres committed:
Update code/modules/reagents/chemistry/reagents/other_reagents.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

mc-oofert commented:
lame title, expected better blood effects of the injury and combat kind

timothymtorres commented:
Updated OP with radiation, dirty, and purity % effects.

timothymtorres committed:
Add disgust and wakeup effects to shower

timothymtorres committed:
Lower the disgust a little

timothymtorres committed:
Increase the drunkness removal

timothymtorres committed:
Add medical tip for showers

timothymtorres commented:
Added a few more effects:
- Dirty showers now generate disgust
- Water TOUCH|VAPOR interactions now heal sleep, unconsciousness, confusion, drowsiness, jitters, dizziness, and drunkenness
- Added shower effects to tips.txt

zoomachina commented:
What does "dirty shower" entail? Does this mean that synthflesh (or any other interesting) showers would give disgust now?

timothymtorres commented:
Anything that isn't +70% regular water

timothymtorres committed:
Merge branch 'master' into blood-shower

timothymtorres committed:
Fix wrong status effect for watery tile

timothymtorres committed:
Fix hot springs status effect

Loafin34 commented:
insane water buff. now for maximum athletics i sleep under a shower to wake up faster

timothymtorres commented:
That's not how athletics works. I designed that skill to only gain xp _while_ a person is sleeping so waking up faster interrupts this process. You are better off resting in good quality conditions (ie. lights off, bedsheet over your mob, sleeping in bed, etc.) to maximize xp gains.

Ghommie commented:
> insane water buff. now for maximum athletics i sleep under a shower to wake up faster

Lol, grindset meme moment.

Wallemations reviewed (approved):
can't remember if I approved this or not, but icons look good

timothymtorres committed:
Refactor optional arguments to the back

timothymtorres commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -129,10 +129,10 @@
 
 /// Applies this reagent to an [/obj]
-/datum/reagent/proc/expose_obj(obj/exposed_obj, reac_volume)
+/datum/reagent/proc/expose_obj(obj/exposed_obj, methods=TOUCH, reac_volume, show_message=TRUE)
```
Applied with https://github.com/tgstation/tgstation/pull/87747/commits/f1761066f7e2f0b38951d35d7c82080f7b2d4b0f

I was following the same pattern as:

`/datum/reagent/proc/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message = TRUE, touch_protection = 0)`

timothymtorres committed:
Apply code review

timothymtorres committed:
Refactor and remove one letter var R for reagent

timothymtorres committed:
Fix spelling typo

timothymtorres committed:
Fix another misspelling

timothymtorres committed:
Bad regex replace mistake

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -282,6 +308,16 @@
+		exposed_mob.adjust_confusion(-2 SECONDS)
+		exposed_mob.adjust_drowsiness(-4 SECONDS)
+		exposed_mob.adjust_jitter(-4 SECONDS)
+		exposed_mob.AdjustSleeping(-150)
```
```suggestion
		exposed_mob.AdjustSleeping(-15 SECONDS)
```

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -282,6 +308,16 @@
+		exposed_mob.adjust_drowsiness(-4 SECONDS)
+		exposed_mob.adjust_jitter(-4 SECONDS)
+		exposed_mob.AdjustSleeping(-150)
+		exposed_mob.AdjustUnconscious(-80)
```
```suggestion
		exposed_mob.AdjustUnconscious(-8 SECONDS)
```

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -126,6 +132,26 @@
+	if(reac_volume < 3)
+		return
+
+	if(methods & (VAPOR|TOUCH))
```
early return

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/87747/commits/0fd7e00ad7a9d8f488b2f13f3df1a45983cd5905

Ghommie commented on code/datums/mood_events/_mood_event.dm:
```diff
@@ -18,11 +18,11 @@
 	var/list/required_job = list()
 
-/datum/mood_event/New(mob/M, ...)
+/datum/mood_event/New(mob/living/M, ...)
```
If you would rename the argument from M to something more descriptive while you're at it. Thank you ^^

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/87747/commits/534fdd5248561a97469bd2da49ef616cf29fefbd

Ghommie commented on code/game/turfs/turf.dm:
```diff
@@ -695,9 +695,9 @@ GLOBAL_LIST_EMPTY(station_turfs)
-	for(var/reagent in reagents)
-		var/datum/reagent/R = reagent
-		. |= R.expose_turf(src, reagents[R])
+	for(var/datum/reagent/reagent in reagents)
```
as anything

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/87747/commits/0fd7e00ad7a9d8f488b2f13f3df1a45983cd5905

timothymtorres committed:
Change numeric argument to use SECONDS define

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

timothymtorres committed:
Change blood to early return for expose_obj

timothymtorres committed:
Merge branch 'blood-shower' of https://github.com/timothymtorres/tgstation into blood-shower

timothymtorres committed:
Optimize reagents loop

timothymtorres committed:
Rename one letter var to something more descriptive
